src/v/storage/segment_index.cc

```diff
@@ -180,7 +180,13 @@ segment_index::find_above_size_bytes(size_t distance) {
         return std::nullopt;
     }
     int i = std::distance(_state.position_index.begin(), it);
-    return translate_index_entry(_state, _state.get_entry(i));
+    std::optional<entry> entry = translate_index_entry(
+      _state, _state.get_entry(i));
+    if (max_offset != model::offset{} && entry->offset > max_offset) {
+        // Fallback to offset based lookup
+        entry = find_nearest(max_offset);
+    }
+    return entry;
```

Comment:
what if method inputs are something like `(distance=100, max_offset=0)`, in this implementation it seems to me the return will be the first index entry or something like that but wouldn't we want a nullopt instead?